tokio-postgres/tests/test/replication.rs

```diff
@@ -0,0 +1,230 @@
+    let slotdesc = rclient
+        .create_physical_replication_slot(slot, false, false)
+        .await
+        .unwrap();
+    assert_eq!(slotdesc.slot_name(), slot);
+    assert_eq!(slotdesc.snapshot_name(), None);
+    assert_eq!(slotdesc.output_plugin(), None);
+
+    let mut physical_stream = rclient
+        .start_physical_replication(None, identify_system.xlogpos(), None)
+        .await
+        .unwrap();
+
+    let _nrows = sclient
+        .execute("insert into test_physical_replication values(1)", &[])
+        .await
+        .unwrap();
+
+    let mut got_xlogdata = false;
+    while let Some(replication_message) = physical_stream.next().await {
+        if let ReplicationMessage::XLogData(_) = replication_message.unwrap() {
+            got_xlogdata = true;
+            break;
+        }
+    }
+
+    assert!(got_xlogdata);
+
+    let response = physical_stream.stop_replication().await.unwrap();
+    assert!(response.is_none());
+
+    // repeat simple command after stream is ended
+    let show_port = rclient.show("port").await.unwrap();
+    assert_eq!(show_port, "5433");
+
+    simple_exec(&sclient, "drop table if exists test_physical_replication").await;
+}
+
+// test for:
+//   - create/drop slot
+//   X standby_status_update
+//   - logical replication
+#[tokio::test]
+async fn logical_replication() {
+    let (sclient, mut rclient) = setup(ReplicationMode::Logical).await;
+
+    simple_exec(&sclient, "drop table if exists test_logical_replication").await;
+    simple_exec(&sclient, "drop publication if exists test_logical_pub").await;
+    simple_exec(&sclient, "create table test_logical_replication(i int)").await;
+    simple_exec(
+        &sclient,
+        "create publication test_logical_pub for table test_logical_replication",
+    )
+    .await;
+
+    let identify_system = rclient.identify_system().await.unwrap();
+    assert_eq!(identify_system.dbname().unwrap(), "postgres");
+
+    let slot = "test_logical_slot";
+    let plugin = "pgoutput";
+    let _ = rclient.drop_replication_slot(slot, false).await.unwrap();
+    let slotdesc = rclient
+        .create_logical_replication_slot(slot, false, plugin, None)
+        .await
+        .unwrap();
+    assert_eq!(slotdesc.slot_name(), slot);
+    assert!(slotdesc.snapshot_name().is_some());
+    assert_eq!(slotdesc.output_plugin(), Some(plugin));
+
+    let xlog_start = identify_system.xlogpos();
+    let options = &vec![
+        ("proto_version", "1"),
+        ("publication_names", "test_logical_pub"),
+    ];
+
+    let mut logical_stream = rclient
+        .start_logical_replication(slot, xlog_start, options)
+        .await
+        .unwrap();
+
+    let _nrows = sclient
+        .execute("insert into test_logical_replication values(1)", &[])
+        .await
+        .unwrap();
+
+    let mut got_begin = false;
+    let mut got_insert = false;
+    let mut got_commit = false;
+    while let Some(replication_message) = logical_stream.next().await {
```

Comment:
@jeff-davis I've been pouring over this PR the past couple of days trying to understand everything and there's one thing I'm not quite connecting. Forgive me, as I'm somewhat new to Rust. Isn't `logical_stream.next()` going to wait indefinitely until there is some copy data to consume from the server? 

The reason I ask is because, if we look at [pg_recvlogical](https://github.com/postgres/postgres/blob/master/src/bin/pg_basebackup/pg_recvlogical.c), we don't want to wait indefinitely on copy data because there may be a need to send standby data to postgres (e.g. if we've waited some long amount of time in between receiving copy data) so that postgres doesn't kill our replication connection. 

Apologies if I'm missing something obvious here.

Reply:
I am not experienced in this subject, but from what I observed while I was [integrating the branch](https://github.com/sfackler/rust-postgres/issues/116#issuecomment-1018486358) by @petrosagg, is that postgres sends out "keepalive" messages every 20 seconds or so.

Interestingly, check [this doc page](https://www.postgresql.org/docs/10/protocol-replication.html) about those keepalive messages; the last bit in the "Primary keepalive message (B)" structure tells the client if postgres detects that the client hasn't communicated in a while, and is nearing the point where it will close the connection.

So what I have [my code do](https://github.com/debate-map/app/blob/afc6467b6c6c961f7bcc7b7f901f0ff5cd79d440/Packages/app-server-rs/src/pgclient.rs#L66) at the moment, is just wait for a keepalive message with that "timeout imminent" flag, and then immediately send a response to the server saying "I am still connected". Note that postgres is actually pretty lenient; it gives another 10 or 15 seconds before it actually closes the connection, which is enough time for this "reactionary" approach to be reliable (during the several hours I've tested it anyway).

From what I can tell, this "notify postgres we're still online" feature is not present in this pull-request's code -- so I guess that's something @petrosagg is thinking can be handled by higher level modules/crates. (it might be worth having at least a warning about this behavior in the base `rust-postgres` library however, eg. if it detects a disconnect, and it can see that it was due to the client not sending any responses during the timeout period)

Reply:
@Venryx unfortunately it is not enough to wait for postgres to send your a keepalive with that bit set in order to send a keepalive, you actually need to proactively be sending them in a timely manner. This is exactly what we do in materialize. 

Check out this code https://github.com/MaterializeInc/materialize/blob/main/src/dataflow/src/source/postgres.rs#L303-L329

and this thread https://www.postgresql.org/message-id/CAMsr+YE2dSfHVr7iEv1GSPZihitWX-PMkD9QALEGcTYa+sdsgg@mail.gmail.com

Reply:
> @Venryx unfortunately it is not enough to wait for postgres to send your a keepalive with that bit set in order to send a keepalive, you actually need to proactively be sending them in a timely manner. This is exactly what we do in materialize.

Ah, interesting; thank you for mentioning. (At my project's current scale, I think it will be a long time before there are transactions of a large enough size for queuing to cause problems with the keepalive responder; but it's still a good thing to fix, to not cause issues for other developers [or myself] down the road.)